Sources/Apodini/Semantic Model Builder/SharedSemanticModelBuilder.swift

```diff
@@ -118,15 +118,70 @@ class SharedSemanticModelBuilder: SemanticModelBuilder {
                 .whenAllSucceed(guardEventLoopFutures, on: request.eventLoop)
                 .flatMap { _ in
                     request.enterRequestContext(with: component) { component in
-                        var response: Encodable = component.handle()
-                        for responseTransformer in responseModifiers {
-                            response = request.enterRequestContext(with: responseTransformer()) { responseTransformer in
-                                responseTransformer.transform(response: response)
-                            }
+                        let response = component.handle()
+                        let promise = request.eventLoop.makePromise(of: Encodable.self)
+                        let visitor = ActionVisitor(request: request,
+                                                    promise: promise,
+                                                    responseModifiers: responseModifiers)
+                        switch response {
+                        case is ApodiniEncodable:
+                            // is an Action
+                            // use double dispatch to access
+                            // wrapped element
+                            // swiftlint:disable:next force_cast
+                            (response as! ApodiniEncodable).accept(visitor)
+                        default:
+                            // not an action
+                            // we can skip the visitor
+                            visitor.visit(encodable: response)
                         }
-                        return request.eventLoop.makeSucceededFuture(response)
+                        return promise.futureResult
                     }
                 }
         }
     }
 }
+
+struct ActionVisitor: ApodiniEncodableVisitor {
+    var request: Request
+    var promise: EventLoopPromise<Encodable>
+    var responseModifiers: [() -> (AnyResponseTransformer)]
+
+    func visit<Element: Encodable>(encodable: Element) {
+        let result = transformResponse(encodable)
+        promise.succeed(result)
+    }
+
+    func visit<Element: Encodable>(action: Action<Element>) {
+        do {
+            switch action {
+            case let .send(element):
+                let result = transformResponse(element)
+                promise.succeed(result)
+            case let .final(element):
+                let result = transformResponse(element)
+                promise.succeed(result)
+                try request.eventLoop.close()
+            case .end:
+                // send back an empty response
+                promise.succeed("")
```

Comment:
I think the exporter should be allowed to handle the `Action` itself. One exporter might be able to close a connection without sending an extra message while others indeed have to send an empty message. I would thus make `ActionVisitor.promise` (and as a consequence the return-value of `RequestHandler`) an `EventLoopFuture<Action<Encodable>>`.

Reply:
I like the idea of delegating the handling of the `Action` to the exporter. This would also keep us from having to make assumptions about how the proper reaction to `.end` looks like.
Unfortunately, `Action<Encodable>` doesn't work (because `Value of protocol type 'Encodable' cannot conform to 'Encodable'` 💩). I used the `AnyEncodable` instead, which makes it a bit uglier but good enough in my opinion.